### PR TITLE
Add directory upload and clear input to `FileInput`

### LINKED ISSUE #13861: [FEATURE] Support directory upload and clearing inputs from Python for `FileInput` widget
### Problem description

I have two requests both related to the `FileInput` widget. 

1) Make it possible to upload directories with FileInput
The HTML element `input` has the property `webkitdirectory`, more information [here](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/webkitdirectory). This property makes it possible to upload a directory. Note that when uploading a directory it will give a warning in the browser for security reasons. 
image

2) Clearing the inputs
I would also like to be able to clear the inputs from the Python side. In my implementation, this is done inelegantly because the parameters are `readonly` (more to show it is possible).

### Feature description

1) Add the `webkitdirectory` property in BokehJS and expose it in Bokeh (I would suggest removing `webkit` from the name)

2) Have a `clear_input` method or similar

### Potential alternatives

None. 

### Additional information

I have made a POC of it in Panel: https://github.com/holoviz/panel/pull/6808/files

But most of it is just coping over the FileInput widgets from Bokeh with minor changes.

### PR DESCRIPTION
**Pull request by hoxbro**
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [x] issues: fixes #13861
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

Unit test: http://127.0.0.1:5777/unit/run?k=directory

Current test script:

``` python
from bokeh.models import FileInput, Button
from bokeh.io import curdoc

file_input = FileInput(directory=True)
button = Button(label="Clear file input")
button.on_click(lambda *e: setattr(file_input, "_clear_input",  file_input._clear_input + 1))

curdoc().add_root(file_input)
curdoc().add_root(button)
```

https://github.com/bokeh/bokeh/assets/19758978/92026137-c4eb-498c-8c9c-c2a303af51bb

### CODE REVIEW

File: bokehjs/src/lib/models/widgets/file_input.ts [hunk 1]
```diff
@@ -103,12 +126,14 @@ export class FileInput extends InputWidget {
...
 
-    this.define<FileInput.Props>(({Bool, Str, List, Or}) => ({
+    this.define<FileInput.Props>(({Bool, Str, List, Or, Int}) => ({
       value:     [ Or(Str, List(Str)), p.unset, {readonly: true} ],
       mime_type: [ Or(Str, List(Str)), p.unset, {readonly: true} ],
       filename:  [ Or(Str, List(Str)), p.unset, {readonly: true} ],
       accept:    [ Or(Str, List(Str)), "" ],
       multiple:  [ Bool, false ],
+      directory: [ Bool, false ],
+      _clear_input: [ Int, 0],
```
Comment (hoxbro) [inline bokehjs/src/lib/models/widgets/file_input.ts]: This is not the final design (I hope). 

I need help figuring out how to tell the input to be cleared from Python to BokehJS. Can an `event` be used?

One of the problems, at least for me, is that `value` is read-only from the Python side.

File: bokehjs/src/lib/models/widgets/file_input.ts [hunk 2]
```diff
@@ -36,22 +36,38 @@ export class FileInputView extends InputWidgetView {
...
     for (const file of files) {
       const data_url = await this._read_file(file)
       const [, mime_type="",, value=""] = data_url.split(/[:;,]/, 4)
 
-      values.push(value)
-      filenames.push(file.name)
-      mime_types.push(mime_type)
+      if (directory) {
+        const ext = file.name.split(".").pop()
+        if ((!is_nullish(accept) && isString(ext)) ? accept.includes(`.${ext}`) : true) {
```
Comment (hoxbro) [inline bokehjs/src/lib/models/widgets/file_input.ts]: This check will likely need some more thought. 

What comes to mind is that it should also check the `mime_type`.

**Review by mattpap — APPROVED** [review]

**Comment by bryevdv** [comment]
@hoxbro did this PR actually add the clear input portion of the original issue, or just the directory upload option? If the clear feature still does not exist via a public API, then I think a new issue just for that part should be opened.

**Comment by hoxbro** [comment]
It was added in an earlier merged pr https://github.com/bokeh/bokeh/pull/13890

**Pull request merged**